jwst/lib/pipe_utils.py

```diff
@@ -0,0 +1,40 @@
+"""Pipeline utilities objects"""
+
+from ..associations.lib.dms_base import TSO_EXP_TYPES
+from ..datamodels import CubeModel
+
+# Model types that typically represent TSO's
+TSO_MODEL_TYPES = (CubeModel,)
+
+
+def is_tso(model):
+    """Is data Time Series Observation data?
+
+    Parameters
+    ----------
+    model: datamodels.DataModel
+        Data to check
+
+    Returns
+    -------
+    is_tso: bool
+       `True` if the model represents TSO data
+    """
+    is_tso = False
+
+    # Check on exposure types
+    try:
+        is_tso = model.meta.exposure.type.lower() in TSO_EXP_TYPES
+    except AttributeError:
+        pass
+
+    # Check on JWST-specific TSOVISIT flag
+    try:
+        is_tso = is_tso or model.meta.observation.tsovisit
+    except AttributeError:
+        pass
+
+    # Check on DataModel types
+    is_tso = is_tso or isinstance(model, TSO_MODEL_TYPES)
```

Comment:
Need to be careful with this one. `TSO_MODEL_TYPES` contains `CubeModel` and while most TSO data may be represented using `CubeModel`, not all `CubeModel` instances are TSO. Coronagraphy exposures, for example, are stored as `CubeModel`, because we need the per-integration data. I think it's safer to stick with defining TSO based on only the meta attributes (EXP_TYPE, TSOVISIT) and not the datamodel instance.

Reply:
Of course if you only ever execute this check within a pipeline like `calwebb_spec2`, then you're safe, because coronagraphy data are image-like and hence go through `calwebb_image2`.

Reply:
👍 on removing the `DataModel` check. Main reason for the review.